### ISSUE TITLE
Sdk 4.6.0 can't connect against 5.0.0-pre1 versions of node

### ISSUE THREAD

**Issue by tenuki** [aeternity/aepp-sdk-js#657]
[api.zip](https://github.com/aeternity/aepp-sdk-js/files/3605413/api.zip)

## Expected Behavior

Sdk allows developer to connect to node v5 versions (-pre1 nor -rc1 from docker/master @ 11/sept).

## Current Behavior
It can't connect. First error display is:

    (node:31818) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'map' of undefined

I've include api document but fails to parse these:

 1. endpoints with no parameters before this, had parameters property = [ ] . 
 2. parameters types are references to parameters referred elsewhere in the api. 

## Possible Solution
There are three problems to achieve this:

1.  node /api description isn't using anymore api call parameter description: parameters: [ ]  when there call has no parameters. This requires to modify:

    function pascalizeParameters (parameters) {
      return parameters.map(o => R.assoc('name', snakeToPascal(o.name), o))
    }

with: 

```diff
@@ -240,7 +247,10 @@ function classifyParameters (parameters) {
  * @return {Object[]} Pascalized parameters
  */
 function pascalizeParameters (parameters) {
-  return parameters.map(o => R.assoc('name', snakeToPascal(o.name), o))
+  if (parameters == null) {
+    return []
+  }
+  return parameters.map(o => R.assoc('name', snakeToPascal(o.name), o))
 }
 
 /**
```

Also:

    const Swagger = stampit({
      init ({ swag = this.swag, axiosConfig }, { stamp }) {
        const { paths, definitions } = swag
        const methods = R.indexBy(
          R.prop('name'),
          R.flatten(
            R.values(
              R.mapObjIndexed(
                (methods, path) => R.values(
                  R.mapObjIndexed((definition, method) => {
                    const op = operation(path, method, definition, definitions, axiosConfig)
                    return op(this, this.urlFor(swag.basePath, definition))
                  }, methods)),
                paths
              )
            )
          )
        )

with: 

```diff
@@ -461,6 +472,9 @@ const Swagger = stampit({
           R.mapObjIndexed(
             (methods, path) => R.values(
               R.mapObjIndexed((definition, method) => {
+                if (definition.parameters === undefined) {
+                  definition.parameters = []
+                }
                 const op = operation(path, method, definition, definitions, axiosConfig)
                 return op(this, this.urlFor(swag.basePath, definition))
               }, methods)),
```

And there is other issue which is that some parameters were replaced with references. Example:
```json
{ 
    "/key-blocks/height/{height}": {
      "get": {
        "tags": ["external", "chain"],
        "description": "Get a key block by height",
        "operationId": "GetKeyBlockByHeight",
        "produces": ["application/json"],
        "parameters": [
          {
            "name": "height",
            "in": "path",
            "description": "The height",
            "required": true,
            "type": "integer",
            "maximum": 18446744073709551615,
            "minimum": 0
          }
        ],
..
```

to:

```json
    "/key-blocks/height/{height}": {
      "get": {
        "tags": ["external", "chain"],
        "operationId": "GetKeyBlockByHeight",
        "description": "Get a key block by height",
        "produces": ["application/json"],
        "parameters": [
          {
            "$ref": "#/parameters/HeightIn"
          }
        ],
```
where `name` isn't directly available.

A quick incomplete workaround for that would be just to ignore those, mainly in the same code above for pascalizeParameters with:

```diff
@@ -233,6 +232,14 @@ function classifyParameters (parameters) {
   }
 }
 
+function safeReplace (anObject) {
+  try {
+    return snakeToPascal(anObject.name)
+  } catch (err) {
+    return null
+  }
+}
+
 /**
  * Convert `name` attributes in `parameters` from snake_case to PascalCase
  * @rtype (parameters: [{name: String}...]) => [{name: String}...]
@@ -240,7 +247,10 @@ function classifyParameters (parameters) {
  * @return {Object[]} Pascalized parameters
  */
 function pascalizeParameters (parameters) {
-  return parameters.map(o => R.assoc('name', snakeToPascal(o.name), o))
+  if (parameters == null) {
+    return []
+  }
+  return parameters.map(o => R.assoc('name', safeReplace(o.name), o))
 }
 
 /**
```

## Steps to Reproduce
1. make sdk 4.6.0 connect to a v5.0.0-pre1 node.

**Comment by tenuki** [comment]
better fix (sorry i've mistaken) : 
```diff
@@ -233,6 +232,17 @@ function classifyParameters (parameters) {
   }
 }
 
+function safeReplace (anObject) {
+  try {
+    return snakeToPascal(anObject.name)
+  } catch (err) {
+    if (anObject != null) {
+      return anObject
+    }
+    return null
+  }
+}
+
 /**
  * Convert `name` attributes in `parameters` from snake_case to PascalCase
  * @rtype (parameters: [{name: String}...]) => [{name: String}...]
@@ -240,7 +250,10 @@ function classifyParameters (parameters) {
  * @return {Object[]} Pascalized parameters
  */
 function pascalizeParameters (parameters) {
-  return parameters.map(o => R.assoc('name', snakeToPascal(o.name), o))
+  if (parameters == null) {
+    return []
+  }
+  return parameters.map(o => R.assoc('name', safeReplace(o), o))
 }
 
 /**
@@ -348,7 +361,7 @@ const operation = (path, method, definition, types, { config, errorHandler } = {
   config = config || {}
   delete config.transformResponse // Prevent of overwriting transform response
   const { operationId, description } = definition
-  const { parameters } = definition
+  let { parameters } = definition
   const name = `${R.head(operationId).toLowerCase()}${R.drop(1, operationId)}`
   const pascalized = pascalizeParameters(parameters)
 
@@ -461,6 +474,9 @@ const Swagger = stampit({
           R.mapObjIndexed(
             (methods, path) => R.values(
               R.mapObjIndexed((definition, method) => {
+                if (definition.parameters === undefined) {
+                  definition.parameters = []
+                }
                 const op = operation(path, method, definition, definitions, axiosConfig)
                 return op(this, this.urlFor(swag.basePath, definition))
               }, methods)),

```

**Comment by nduchak** [comment]
Hello, 
The sdk 4.6.0 is support node 5.0.0-rc.1. How fo you run your node(using docker-hub or from repo)?

**Comment by tenuki** [comment]
> The sdk 4.6.0 is support node 5.0.0-rc.1. How fo you run your node(using docker-hub or from repo)?

From docker-hub installed master, yesterday.  Thanks @nduchak !

**Comment by nduchak** [comment]
You are welcome)

**Comment by tenuki** [comment]
But sdk 4.6.0 failed with that. Where is it fixed? Thanks!

**Comment by nduchak** [comment]
It will be fixed soon

**Comment by tenuki** [comment]
They say it is already fixed, but I was using incompatible versions 4.6.0/4.7.0 against v5.0.0-rc1-branch.

I think this *should* be closed.
